metals/src/main/scala/scala/meta/internal/metals/testProvider/TestDiscovery.scala

```diff
@@ -0,0 +1,46 @@
+package scala.meta.internal.metals.testProvider
+
+import javax.annotation.Nullable
+
+import scala.meta.internal.metals.MetalsEnrichments._
+
+import org.eclipse.{lsp4j => l}
+
+final case class TestDiscovery(
+    targetName: String,
+    targetUri: String,
+    discovered: java.util.List[TestDiscoveryResult]
+)
+
+sealed trait TestDiscoveryResult {
+  def nonEmpty: Boolean = this match {
+    case p: TestDiscoveryResult.Package => p.children.asScala.nonEmpty
+    case _: TestDiscoveryResult.TestClass => true
+  }
+}
+
+object TestDiscoveryResult {
+
+  final class Package private (
+      val kind: String,
+      val prefix: String,
+      val children: java.util.List[TestDiscoveryResult]
+  ) extends TestDiscoveryResult
+
+  object Package {
+    def apply(prefix: String, children: java.util.List[TestDiscoveryResult]) =
+      new Package("package", prefix, children)
+  }
+
+  final class TestClass private (
+      val kind: String,
+      val fullyQualifiedName: String,
+      val className: String,
+      @Nullable val location: l.Location
+  ) extends TestDiscoveryResult
+
+  object TestClass {
+    def apply(pkg: String, testName: String, @Nullable location: l.Location) =
+      new TestClass("class", pkg, testName, location)
+  }
+}
```

Comment:
It's the best way to model discriminated union type which I discovered. These classes could be `case` classes but then the proper value of `kind` won't be enforced because of generated apply.